// @esri/arcgis-rest-portal/dist/node/items/content.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var arcgis_rest_request_1 = require("@esri/arcgis-rest-request");
var get_portal_url_1 = require("../util/get-portal-url");
var helpers_1 = require("./helpers");
/**
 * ```js
 * import { getUserContent } from "@esri/arcgis-rest-portal";
 * //
 * getUserContent({
 *    owner: 'geemike',
 *    folderId: 'bao7',
 *    start: 1,
 *    num: 20,
 *    authentication
 * })
 * ```
 * Returns a listing of the user's content. If the `username` is not supplied, it defaults to the username of the authenticated user. If `start` is not specificed it defaults to the first page.
 * If the `num` is not supplied it is defaulted to 10. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/user-content.htm) for more information.
 *
 * @param requestOptions - Options for the request
 * @returns A Promise<IUserContentResponse>
 */
exports.getUserContent = function (requestOptions) {
    var folder = requestOptions.folderId, _a = requestOptions.start, start = _a === void 0 ? 1 : _a, _b = requestOptions.num, num = _b === void 0 ? 10 : _b, authentication = requestOptions.authentication;
    var suffix = folder ? "/" + folder : "";
    return helpers_1.determineOwner(requestOptions)
        .then(function (owner) { return get_portal_url_1.getPortalUrl(requestOptions) + "/content/users/" + owner + suffix; })
        .then(function (url) { return arcgis_rest_request_1.request(url, {
        httpMethod: "GET",
        authentication: authentication,
        params: {
            start: start,
            num: num,
        },
    }); });
};
