// @esri/arcgis-rest-portal/dist/node/items/export.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var arcgis_rest_request_1 = require("@esri/arcgis-rest-request");
var helpers_1 = require("./helpers");
var get_portal_url_1 = require("../util/get-portal-url");
/**
 * ```js
 * import { exportItem } from "@esri/arcgis-rest-portal";
 * //
 * exportItem({
 *      id: '3daf',
 *      owner: 'geemike',
 *      exportFormat: 'CSV',
 *      exportParameters: {
 *        layers: [
 *          { id: 0 },
 *          { id: 1, where: 'POP1999 > 100000' }
 *        ]
 *      },
 *      authentication,
 *    })
 * ```
 * Exports an item from the portal. See the [REST Documentation](https://developers.arcgis.com/rest/users-groups-and-items/export-item.htm) for more information.
 *
 * @param requestOptions - Options for the request
 * @returns A Promise<IExportItemResponse>
 */
exports.exportItem = function (requestOptions) {
    var authentication = requestOptions.authentication, itemId = requestOptions.id, title = requestOptions.title, exportFormat = requestOptions.exportFormat, exportParameters = requestOptions.exportParameters;
    return helpers_1.determineOwner(requestOptions)
        .then(function (owner) { return get_portal_url_1.getPortalUrl(requestOptions) + "/content/users/" + owner + "/export"; })
        .then(function (url) { return arcgis_rest_request_1.request(url, {
        httpMethod: 'POST',
        authentication: authentication,
        params: {
            itemId: itemId,
            title: title,
            exportFormat: exportFormat,
            exportParameters: exportParameters
        }
    }); });
};
